### DISCUSSION TITLE
incomplete attributes in metadata structs

### DISCUSSION THREAD

**Discussion by eladyn** [librespot-org/librespot#1054] [Feature Requests]
Recently, there have been several discussions about the current structs in the `librespot_metadata` crate. Since I have been involved in some of them myself, I thought I'd just write down the issue and some of my thoughts in a central place, as a starting point for future discussion.

As far as I can see, two problems have been popping up with the current implementation:

## Problem 1: missing attributes in API responses

Depending on the way a specific metadata struct has been instantiated, the degree to which attributes do actually have sensible values varies widely, and rarely all fields are covered.

One such example is `metadata::album::Album` struct:

- A request to `/metadata/4/album/<id>` returns quite some info; some attributes are still missing, but many are there:
  <details>
  
  ```
  gid: "\257^\336Dw?Cv\203\211>\0304X\272~"
    name: "Ten Summoner\'s Tales"
    artist {
      gid: "\035pL,,PK\266\264s\246\275\343\217\265\312"
      name: "Sting"
    }
    type: ALBUM
    label: "A&M"
    date {
      year: 1993
      month: 3
      day: 9
    }
    popularity: 63
    external_id {
      type: "upc"
      id: "00602567834496"
    }
    disc {
      number: 1
      track {
        gid: "\002\353\324\215*\222H\233\203\337\020\344cz\250\322"
      }
      track {
        gid: "s9\246 \270\271G\325\222\376\206l$\231\211\024"
      }
      track {
        gid: "\027T\360h\253]KF\251P\035i\356\201\234\242"
      }
      track {
        gid: "\216\354m\n/\337K\354\232\360\2758\222?\201\244"
      }
      track {
        gid: "\244\247/r\310\236@,\2264\205v\336\274Il"
      }
      track {
        gid: "`\244P\330\253\306H\001\260uO8\032\304\312r"
      }
      track {
        gid: "\233\355p@K\355Am\243x\314\240zdOE"
      }
      track {
        gid: "\335\203\221o\032\353D\241\247\\T\343x\317\304\360"
      }
      track {
        gid: "\223i\021\331\313\265HK\240\007t\220DpJ\343"
      }
      track {
        gid: "\331\255\003u\007/L\364\234\324!@\347\207\031\211"
      }
      track {
        gid: "\347\006KNJqJY\207\215\273\005 \3209\303"
      }
      track {
        gid: "\360\230\022\242B\261B\033\200\315\331ufK\243?"
      }
    }
    copyright {
      type: C
      text: "\302\251 1993 UMG Recordings, Inc."
    }
    copyright {
      type: P
      text: "An A&M Records release; \342\204\227 1993 UMG Recordings, Inc."
    }
    cover_group {
      image {
        file_id: "\253gam\000\000\036\002e;\021\r\225`\353\026V\364\305\203"
        size: DEFAULT
        width: 300
        height: 300
      }
      image {
        file_id: "\253gam\000\000HQe;\021\r\225`\353\026V\364\305\203"
        size: SMALL
        width: 64
        height: 64
      }
      image {
        file_id: "\253gam\000\000\262se;\021\r\225`\353\026V\364\305\203"
        size: LARGE
        width: 640
        height: 640
      }
    }
    original_title: "Ten Summoner\'s Tales"
    ```
  </details>
- When requesting a track via `/metadata/4/track/<id>` however, which contains information about its containing album as well, we receive less information:
  <details>

  ```
    album {
      gid: "\257^\336Dw?Cv\203\211>\0304X\272~"
      name: "Ten Summoner\'s Tales"
      artist {
        gid: "\035pL,,PK\266\264s\246\275\343\217\265\312"
        name: "Sting"
      }
      label: "A&M"
      date {
        year: 1993
        month: 3
        day: 9
      }
      cover_group {
        image {
          file_id: "\253gam\000\000\036\002e;\021\r\225`\353\026V\364\305\203"
          size: DEFAULT
          width: 300
          height: 300
        }
        image {
          file_id: "\253gam\000\000HQe;\021\r\225`\353\026V\364\305\203"
          size: SMALL
          width: 64
          height: 64
        }
        image {
          file_id: "\253gam\000\000\262se;\021\r\225`\353\026V\364\305\203"
          size: LARGE
          width: 640
          height: 640
        }
      }
    }
    ```
    
    </details>

The issue here is now that both are internally parsed with the same `.proto` definitions (`protocol::metadata::Album`) and end up in the same higher-level struct from the metadata crate (`metadata::album::Album`). In the latter, most attributes are filled with default values, as you can see below (this is the album data from the requested track parsed into `metadata::album::Album`):

<details>

```
Album {
    id: SpotifyId(
        "spotify:album:5kV0KBXfELibs6qQJLmOtg",
    ),
    name: "Ten Summoner's Tales",
    artists: Artists(
        [
            Artist {
                id: SpotifyId(
                    "spotify:artist:0Ty63ceoRnnJKVEYP0VQpk",
                ),
                name: "Sting",
                popularity: 0,
                top_tracks: CountryTopTracks(
                    [],
                ),
                albums: AlbumGroups(
                    [],
                ),
                singles: AlbumGroups(
                    [],
                ),
                compilations: AlbumGroups(
                    [],
                ),
                appears_on_albums: AlbumGroups(
                    [],
                ),
                genre: [],
                external_ids: ExternalIds(
                    [],
                ),
                portraits: Images(
                    [],
                ),
                biographies: Biographies(
                    [],
                ),
                activity_periods: ActivityPeriods(
                    [],
                ),
                restrictions: Restrictions(
                    [],
                ),
                related: Artists(
                    [],
                ),
                is_portrait_album_cover: false,
                portrait_group: Images(
                    [],
                ),
                sales_periods: SalePeriods(
                    [],
                ),
                availabilities: Availabilities(
                    [],
                ),
            },
        ],
    ),
    album_type: ALBUM,
    label: "A&M",
    date: Date(
        OffsetDateTime {
            utc_datetime: PrimitiveDateTime {
                date: Date {
                    year: 1993,
                    ordinal: 68,
                },
                time: Time {
                    hour: 0,
                    minute: 0,
                    second: 0,
                    nanosecond: 0,
                },
            },
            offset: UtcOffset {
                hours: 0,
                minutes: 0,
                seconds: 0,
            },
        },
    ),
    popularity: 0,
    genres: [],
    covers: Images(
        [
            Image {
                id: FileId(
                    Ok(
                        "ab67616d00001e02653b110d9560eb1656f4c583",
                    ),
                ),
                size: DEFAULT,
                width: 300,
                height: 300,
            },
            Image {
                id: FileId(
                    Ok(
                        "ab67616d00004851653b110d9560eb1656f4c583",
                    ),
                ),
                size: SMALL,
                width: 64,
                height: 64,
            },
            Image {
                id: FileId(
                    Ok(
                        "ab67616d0000b273653b110d9560eb1656f4c583",
                    ),
                ),
                size: LARGE,
                width: 640,
                height: 640,
            },
        ],
    ),
    external_ids: ExternalIds(
        [],
    ),
    discs: Discs(
        [],
    ),
    reviews: [],
    copyrights: Copyrights(
        [],
    ),
    restrictions: Restrictions(
        [],
    ),
    related: Albums(
        [],
    ),
    sale_periods: SalePeriods(
        [],
    ),
    cover_group: Images(
        [
            Image {
                id: FileId(
                    Ok(
                        "ab67616d00001e02653b110d9560eb1656f4c583",
                    ),
                ),
                size: DEFAULT,
                width: 300,
                height: 300,
            },
            Image {
                id: FileId(
                    Ok(
                        "ab67616d00004851653b110d9560eb1656f4c583",
                    ),
                ),
                size: SMALL,
                width: 64,
                height: 64,
            },
            Image {
                id: FileId(
                    Ok(
                        "ab67616d0000b273653b110d9560eb1656f4c583",
                    ),
                ),
                size: LARGE,
                width: 640,
                height: 640,
            },
        ],
    ),
    original_title: "",
    version_title: "",
    type_str: "",
    availability: Availabilities(
        [],
    ),
}
```

</details>

This causes several problems:

- Users are left with a lot of uncertainty, what data they can rely on to be existent.
- Parsing errors due to some attributes failing to initialize from the default values. See https://github.com/librespot-org/librespot/pull/1053#discussion_r956760127 and https://github.com/librespot-org/librespot/pull/1019#issue-1289423737 for examples.

## Problem 2: Large struct sizes

As discovered [here](https://github.com/librespot-org/librespot/pull/1053#discussion_r956778327), e.g. `metadata::track::Track` is almost 1 KB on the stack, without all the possible data that a populated `Track` struct would need additionally on the heap.

Some data:
<details>

```
size of librespot_metadata::track::Track is 944 bytes
size of librespot_metadata::Album is 456 bytes
size of librespot_metadata::Episode is 464 bytes
size of librespot_metadata::Show is 296 bytes
size of librespot_core::SpotifyId is 24 bytes
```

</details>

This is, of course, not generally a bad thing, it only becomes one, when most of these fields aren't actually filled with sensible data and still take up all the space.

## Solution Idea 1: `Option<AllTheThings>`

As proposed by @roderickvd [here](https://github.com/librespot-org/librespot/pull/1019#issuecomment-1170782142) (and apparently already discussed somewhere else before), all `struct` attributes that are optional in the protobuf definitions or appear to be sometimes available and sometimes not could be made `Option<T>`s.

This would solve several pain points with "Problem 1". However, the structs would still require as much space as before and some new problems might arise: Code that knows of the existence of optional fields now has to either `.unwrap()` them or `match` them. The former can lead to surprising breakage, when Spotify changes the data they sent and the latter would introduce quite some boilerplate in library *and* user code.

## Solution Idea 2: `struct`s for everyone and everything

As the name size, this proposal would mean to add separate high-level (as in: contained in the `metadata` crate) structs for each and every situation that returns different attributes.

Something similar does already exist in [`ArtistWithRole`](https://github.com/librespot-org/librespot/blob/16dbade51665667ecf768f21099afc648e0b0176/metadata/src/artist.rs#L59-L64), which is separate from `Artist`. In this case though, the underlying protobuf definition is also different (`protocol::metadata::ArtistWithRole`).

Where it is done exactly like that is the [`rspotify`](https://crates.io/crates/rspotify) crate, e.g. for different [album representations](https://docs.rs/rspotify-model/0.11.5/rspotify_model/album/index.html).

While this might be the ideal solution from a library user's point of view, it potentially comes with a heavy burden for `librespot` developers.

- In order to create those high-level wrapper structs, all endpoints that return those wrapper structs have to be watched and corresponding structs need to be created.
- When Spotify adds or removes attributes that they send, those changes need to be first noticed and the structs be adapted accordingly.
   Although in contrast to the `.unwrap()` in "solution 1", the metadata would not break immediately in most cases, since they are still provided with defaults from the protobufs.

## Solution Idea 3: `enum`s

I haven't yet thought about it more, but @dnlmlr also proposed something with regards to this issue. Since I'm not sure I understand it well enough to write down the details here myself, here's the link with his explanation: https://github.com/librespot-org/librespot/pull/1018#issuecomment-1201006698

---

I hope that this discussion is somehow useful to bundle the various talks into one single place. Looking forward to hearing your thoughts on the matter!

**Comment by dnlmlr** [comment]
This is indeed an issue with no obvious good solution. I think the fact that the same protobuf definitions are used for different responses, containing different data is not great, but I guess that is something that every implementation of the Spotify API will have to deal with.

In general I personally believe that forcing a check is better than silently using a default value for missing fields. When a programmer has to check if a value is present, they can decide to use a default value themselve or take another action. In contrast to that, a default value can be indistinguishable from a real value (e.g. default `0` vs actual value `0`). That takes away control and could lead to the use of *wrong* (default) values, thinking the values were provided by the API. To manually validate these values, an allowed range would have to be known for each field and even then there might be valid values that are the same as the default.
I also think that this would follow the spirit of the rust programming language that provides the `Option` type exactly to represent optional values. To just get the default value in case of `None` there is `Option::unwrap_or_default()` which would result in the same behavior as just using default values in the first place.
That being said, I of course see how it is annoying to `unwrap` every time you want to use a field.

The issue of a program breaking when spotify changes data is not a big issue in my opinion. If not properly prepared, the program will *stop to work the way it was intended* in any case, but with explicit checks you will know about it or even be able to prevent it. In the other case it will just continue to go on with default values that may or may not work. So with `Option`s in a way you will be forced to explicitely decide what to do when encountering missing data: Crash (`unwrap`), use default values (`unwrap_or_default`), or properly handle it (`match`, and others). Instead of just leaving it up to *chance* (not really but pretty much).

## Further explanation of Solution 3
The idea of this proposal was to create an `enum` for each type that represents two states: 
- A minimal subset of values that are present in all responses for a given type
- The full data as it can be requested from the dedicated endpoint (this would be the current metadata types)

This could look something like this (example for `Artist`):
```rust
enum LazyArtist {
  Minimal {
    id: SpotifyId,
    name: Option<String>,
  },
  Full(Box<Artist>),
}
```
Performing a query to the endpoint dedicated to a given type, would always return the `Full` variant. In probably all other cases the `Minimal` variant would be returned. For example when making a request to `/metadata/4/track/<id>`, the album would be the `Minimal` variant. This also mitigates the possible issue of having recursive types (album containing artist, containing albums, containing artist, ...).
To get the `Full` variant from a `Minimal` variant, a `load` function could be used, that will perform a dedicated query to the endpoint for this type. The variant would then be switched to `Full` and be filled with the information from that new query.

This does in my opinion provide a somewhat cleaner Api, but it of course has several disadvantages.
1. What data would exactly be included in the minimal variant? The `id` will be available in every occurrence of a type, but everything else might be optional (if not now, it could be through changes by spotify).
2. If different endpoints return different incomplete versions, we an issue similar to what should be solved in the first place. Should the `Minimal` variant still contain `Optional` values for fields that are present in some endpoint, but not in others? Alternatively these values could just be ignored, which would then require a `load` call to get the data. The `load` would have to perform a full API request, even though a particular datapoint might have already been sent.
3. The `enum` still has to be `match`ed on in a lot of cases, which will create more boilerplate again. This can be improved by implementing functions like `get_id` on the `enum` that will return the `id` without the need to `match`.



While Solution 3 was my proposal, I am really not sure which of the suggested solutions would be the best. All of them have advantages and disadvantages, which makes a decision very hard.

**Reply by eladyn** [reply]
> In general I personally believe that forcing a check is better than silently using a default value for missing fields. When a programmer has to check if a value is present, they can decide to use a default value themselve or take another action. In contrast to that, a default value can be indistinguishable from a real value (e.g. default `0` vs actual value `0`). That takes away control and could lead to the use of _wrong_ (default) values, thinking the values were provided by the API. To manually validate these values, an allowed range would have to be known for each field and even then there might be valid values that are the same as the default. I also think that this would follow the spirit of the rust programming language that provides the `Option` type exactly to represent optional values. To just get the default value in case of `None` there is `Option::unwrap_or_default()` which would result in the same behavior as just using default values in the first place. That being said, I of course see how it is annoying to `unwrap` every time you want to use a field.
> 
> The issue of a program breaking when spotify changes data is not a big issue in my opinion. If not properly prepared, the program will _stop to work the way it was intended_ in any case, but with explicit checks you will know about it or even be able to prevent it. In the other case it will just continue to go on with default values that may or may not work. So with `Option`s in a way you will be forced to explicitely decide what to do when encountering missing data: Crash (`unwrap`), use default values (`unwrap_or_default`), or properly handle it (`match`, and others). Instead of just leaving it up to _chance_ (not really but pretty much).

Fully agree with you on this!

> ## Further explanation of Solution 3

Thanks for elaborating.

>     2. If different endpoints return different incomplete versions, we an issue similar to what should be solved in the first place. Should the `Minimal` variant still contain `Optional` values for fields that are present in some endpoint, but not in others? Alternatively these values could just be ignored, which would then require a `load` call to get the data. The `load` would have to perform a full API request, even though a particular datapoint might have already been sent.

When you take the `Artist` as an example. At least from querying the `/metadata/4/{artist,track,album}` endpoints, it seems that it comes in three variants:

```rust
// used for: artists of an album (`Album::artists`), artists of a track (`Track::artists`), artists of an album of a track (`Track::album::artists`)
struct MinimalArtist {
    id: SpotifyId,
    name: String,
}

// used for: related artists (`Artist::related`)
struct RelatedArtist {
    id: SpotifyId,
    name: String,
    activity_periods: ActivityPeriods,
    portrait_group: Images,
}

// used for: artists (`Artist`)
struct Artist {
    // same as the current definition
}
```

So, implementing your solution here would mean:

```rust
// either
enum LazyArtist {
    Minimal {
        id: SpotifyId,
        name: String,
        activity_periods: Option<ActivityPeriods>, // do we need Option here, otherwise it would be an empty `Vec`?
        portrait_group: Option<Images>, // same question as above 
    },
    Full(Box<Artist>),
}

// or
enum LazyArtist {
    Minimal {
        id: SpotifyId,
        name: String,
    },
    Related {
        id: SpotifyId,
        name: String,
        activity_periods: ActivityPeriods,
        portrait_group: Images,
    },
    Full(Box<Artist>),
}
```

>     3. The `enum` still has to be `match`ed on in a lot of cases, which will create more boilerplate again. This can be improved by implementing functions like `get_id` on the `enum` that will return the `id` without the need to `match`.

This is the main problem, I see with this proposal:

A benefit I can imagine of having the enums in place is, it allows us to do things like the following:

```rust
let mut track = Track::get(&session, &id).await.unwrap();
track.artist = LazyArtist::Full(Box::new(track.artist.load(&session).await.unwrap()));
// now we have a Track with a fully loaded Artist that we can pass around

// However, when using this, we still need to do:
let artist = match track.artist {
    Artist::Full(artist) => artist,
    _ => unreachable!(),
};
// which is obviously not that great, since it can lead to panics due to developer oversights
```

Do you have a better example, where this `enum` structure is useful? Otherwise, I think, we could also apply your idea with the `load` function to Solution 2.

> While Solution 3 was my proposal, I am really not sure which of the suggested solutions would be the best. All of them have advantages and disadvantages, which makes a decision very hard.

Right. :)

**Reply by dnlmlr** [reply]
Your interpretation of solution 3 is pretty much accurate, except the the `load` function.

> A benefit I can imagine of having the enums in place is, it allows us to do things like the following:
> ... (code) ...

This would be at least a little bit better when using the hypothetical `load` how I imagined it. What I thought of was that `load` would switch out the `enum` variant in place if it is not already the full variant. 
So the first part of the code would look like this, which is a bit less verbose and complicated:

```rust
let mut track = Track::get(&session, &id).await.unwrap();
// The id is in both minimal and full, so it can be accessed with a convenicence function without matching (name as well I guess)
let artist_id = track.artist.get_id();
track.artist.load(&session).await.unwrap();
// now we have a Track with a fully loaded Artist that we can pass around
```

The second part would stay the same and still remain a problem. But while writing this, I realized that the load and match could be combined to make it a bit more friendly.

```rust
let mut track = Track::get(&session, &id).await.unwrap();
// This loads the full artist variant inside the track and returns a Clone or Reference to the inner *full* artist
let full_artist = track.artist.load(&session).await.unwrap();

// The full_artist is the actual full artist, so no matches needed to access stuff that is not in minimal
println!("{:#?}", full_artist.cds);

// The track has also switched to the full artist variant, so this still works
let also_full_artist = match track.artist {
    Artist::Full(artist) => artist,
    _ => unreachable!(),
};
```

> Otherwise, I think, we could also apply your idea with the load function to Solution 2.

I guess the `load` with both of our interpretations could be used in either Solution 1 or Solution 2.
Your interpretation with `load` returning the full variant (kind of a shothand for querying via the id) could definitely be used in Solution 2 if wanted.

My own interpretation could in theory be used with Solution 1 to load all the `None` fields. This would still be problematic since you still need all the `unwraps` and the `load` function would need to always do the query, or check all fields as it doesn't know if it is completely filled. In the Solution 3 it can just match against `self` and do nothing if it is the full variant already.

**Comment by roderickvd** [comment]
Thanks so much for this most excellent write-up and subsequent replies.

Just to entertain my earlier thought (without saying I want to steer in that direction per se) could we not provide some macros to make life easy on us (so doing a `has_field()` check then setting an `Option`) as well as to library users (maybe transforming an "`Option`ed" struct into one with all values `unwrap_or_default()`ed should the user choose not to care)?

**Reply by eladyn** [reply]
You're welcome. :)

> could we not provide some macros to make life easy on us (so doing a `has_field()` check then setting an `Option`) 

In some instances, that should be quite easy. The [`Album` struct](https://docs.rs/librespot-protocol/latest/librespot_protocol/metadata/struct.Album.html) for example exposes its fields as `SingularPtrField`s, which contains methods like [`into_option`](https://docs.rs/protobuf/2.27.1/protobuf/struct.SingularPtrField.html). Other instances like [`AudioFile`](https://docs.rs/librespot-protocol/latest/librespot_protocol/metadata/struct.AudioFile.html) have their fields private for whatever reason, and would as such require a macro. But that shouldn't be too complicated.

> to library users (maybe transforming an "Optioned" struct into one with all values unwrap_or_default()ed should the user choose not to care)

The problem here is that

1. the macro would need to create two structs for the same thing, one with `Option`s, the other one without. I'm not too experienced with macro writing, but I imagine this as at least not trivial.
2. additionally, there are currently some data structures that have no sensible defaults, such as the quite integral `SpotifyId` type. This could be changed, but in the process we would lose the property that those types always contain valid invariants, one IMO very useful property of most Rust types.

**Reply by kingosticks** [reply]
> 2. additionally, there are currently some data structures that have no sensible defaults, such as the quite integral SpotifyId type. 

In this case what's wrong with default value `(0, SpotifyItemType::Unknown)`?

**Reply by eladyn** [reply]
Yeah, that would probably work in this case. `to_base62()` and `to_base16()` would both return an `Err` in this case, so I guess there's no way such a value could actually do harm.
